spec/formatting.md

```diff
@@ -18,3 +18,38 @@ the local variable takes precedence.
 
 It is an error for a local variable definition to
 refer to a local variable that's defined after it in the message.
+
+## Bidirectional Isolation
+
+Formatted messages may be bidirectional,
+i.e. contain a mix of left-to-right and right-to-left content.
+In all output formats,
+such parts should be isolated from each other so that
+the directionality of one part of the message does not affect
+the presentation of the rest of the message.
+
+Implementations may provide one or more strategies for isolation,
+but one such strategy must behave as follows,
+when formatting the message as a string:
+
+Considering a "formatted part" to be the formatted string representation of a Placeholder,
+when the formatted part has a direction different from the message as a whole
+or when its directionality is not known,
+the formatted string representation of the message must
+prefix the formatted part with one of the following isolate characters:
+
+- U+2066 LEFT-TO-RIGHT ISOLATE if the formatted part is known to have LTR directionality,
+- U+2067 RIGHT-TO-LEFT ISOLATE if the formatted part is known to have RTL directionality, or
+- U+2068 FIRST STRONG ISOLATE if the formatted part's directionality is not known.
+
+In all cases where a formatted part is prefixed with such an isolate character,
+it must be postfixed with a corresponding U+2069 POP DIRECTIONAL ISOLATE character.
+
+An implementation must define its own method for
+determining the directionality of a formatted part,
+relying either on metadata attached to the formatted part,
+or on introspection of its contents.
```

Comment:
I would add one more callout here. Many of the formatted parts are produced by a formatter with a given locale. APIs exist in at least a few platforms for determining the default base direction based on locale. So perhaps:

> An implementation _must_ determine the base paragraph direction for each formatted part. These might include a metadata field/method, introspection of the formatted part using UBA \[UAX9\], or from the locale of the placeholder's formatter (for example, `Intl.Locale`'s [textInfo](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/Locale/textInfo) in JavaScript). Formatters are responsible for using whatever mechanism is provided by the implementation to provide appropriate directional information for each formatted part.